### PR TITLE
Canonical isomorph and automorphism group

### PR DESCRIPTION
**Pull request by kris-brown**
This PR incorporates the interface with `nauty.c` from CSetAutomorphisms.jl (see https://github.com/AlgebraicJulia/ACSets.jl/issues/3). Rather than relying on a path to the `dreadnaut` executable (a high-level text-based interface for `nauty`), we use the path provided by [nauty_jll](https://github.com/JuliaPackaging/Yggdrasil). 

Nauty operates on undirected graphs, so a big portion of the code involves encoding/decoding ACSets as undirected graphs. The result of calling nauty on an ACSet is a `CanonicalCSet`, which is currently the only subtype of the abstract type `NautyRes` (in general, we will want canonical forms of diagrams of ACSets, etc.). For any `NautyRes` there is an associated hash, orbit information, canonical isomorph, an iso from the input into that canonical isomorph, and a set of generators of the automorphism group. There is also code here which computes the full automorphism group given the generators.

Because ACSetTransformations belong in Catlab rather than ACSets.jl, the notion of a map between ACSets is a `Dict{Symbol, Vector{Int}}`.

Future improvements
- extend this to work with VarACSets
- use the `nauty` function directly rather than producing a string input for `dreadnaut` and parsing the output string.
- Use directed graph rather than undirected graph representation for nauty input (right now, it is inexplicably slow)

### CODE REVIEW

**Comment by kris-brown** [comment]
Seems like Windows is [not a supported platform](https://github.com/JuliaPackaging/Yggdrasil/blob/master/N/nauty/build_tarballs.jl#L51) for `nauty_jll` after all, so I skip those tests for Windows.

**Comment by kris-brown** [comment]
This feature was implicitly requested by Paul Dancstep who was interested in a project related to enumerating slices of graphs up to isomorphism.

**Review by KevinArlin — CHANGES_REQUESTED** [review]

File: src/NautyInterface.jl [hunk 1]
```diff
@@ -0,0 +1,434 @@
...
+        ". f = [$(join([join(c.-1, ",") for c in colorsarray],"|"))]",
+        "c d x b z o"], " ")
+end
+
+"""
+Action of a permutation on a ACSet, X. Results in a new ACSet, Y, and a map X->Y
+"""
+function apply(X::ACSet, p::CPerm)
+  S = acset_schema(X)
+  cd = deepcopy(X)
```
[the remark this answers is not in the document]
Comment (kris-brown) [reply src/NautyInterface.jl:354]: hm may that could be an Issue? I guess that would be something of the right shape but all the homs and attrs are undefined?

File: src/NautyInterface.jl [hunk 2]
```diff
@@ -0,0 +1,434 @@
...
+  for h in homs(S; just_names=true)
+    σs, σt = [collect(x) for x in p.homs[h]]
+    σti = invperm(collect(σt))
+    f = collect(X[h])
```
[the remark this answers is not in the document]
Comment (kris-brown) [reply src/NautyInterface.jl]: yeah the issue is that the raw nauty result seems to specify an iso in the other direction, whereas I think the natural thing (viewing the canonical morphism as something which takes you *to* a normal form) is to have the inverse. 

A solution could be to `inv` the the thing right upon parsing it from nauty output and then `inv` everything here.

File: src/NautyInterface.jl [hunk 3]
```diff
@@ -0,0 +1,434 @@
...
+
+function add_edges!(u::UnDiGraph, sts::Vector{Tuple{Int,Int}})::Vector{Int}
+  map(sts) do (s, t)
+    add_edge!(u, s, t) 
+  end 
+end
+
+add_edges!(u::UnDiGraph, ss::Vector{Int}, ts::Vector{Int})::Vector{Int} =
+  add_edges!(u, collect(zip(ss, ts)))
+
```
[the remark this answers is not in the document]
Comment (kris-brown) [reply src/NautyInterface.jl:190]: Hm would that only help performance if we were to do actual linear algebra with it? My understanding is that the matrix only exists to be converted into a string for `dreadnaut` input.

File: src/NautyInterface.jl [hunk 4]
```diff
@@ -0,0 +1,434 @@
...
+function to_mat(u::UnDiGraph)::Matrix{Bool}
+  mat = zeros(Bool, (u.n,u.n))
+  for (s, t) in zip(u.src, u.tgt)
+    mat[s, t] = mat[t, s] = true
+  end
+  mat
+end
+
```
[the remark this answers is not in the document]
Comment (kris-brown) [reply src/NautyInterface.jl:198]: haha I don't think so! But `prime` seemed appropriate for "rename to be slightly different in a way that probably doesn't clash with any other name". But this makes me now think that things will break if someone actually has a hom `src` and a hom `_src`, so if it sounds good I can replace with:

`prime(hom_name::Symbol) = Symbol("$(hom_name)_$(hash(hom_name))")`

(I'd use `gensym` but this needs to be deterministic)

File: src/NautyInterface.jl [hunk 5]
```diff
@@ -0,0 +1,434 @@
...
+  Dict([k=>Permutation(v) for (k,v) in collect(canon2)])
+end
+
+compose(f::Vector{Int}, g::Vector{Int}) = g[f]
+compose(xs::Vector{Int}...) = foldl(compose, xs)
+
+"""
+Because the hom/attr perms really are just given by the perms on their
+domain/codomain, we don't actually need to track anything other than 'obs'. TODO
+"""
```
[the remark this answers is not in the document]
Comment (kris-brown) [reply src/NautyInterface.jl]: deleted the whole `struct`

File: src/NautyInterface.jl [hunk 6]
```diff
@@ -0,0 +1,434 @@
...
+      append!(colorsarray, [[elem] for elem in oinds[c]])
+    end
+  end
+  colorsarray
+end
+
+"""
+Create partition of flattened indices (all parts altogether) to dict with 
+indices for each distinct object, e.g. {V => 1:8, E => 9:13, ...}
+"""
```
[the remark this answers is not in the document]
Comment (kris-brown) [reply src/NautyInterface.jl:269]: made the docstring more complete!

File: src/NautyInterface.jl (hunk 3 above)
[the remark this answers is not in the document]
Comment (KevinArlin) [reply src/NautyInterface.jl:190]: Well you only allocate ~half the space to a symmetric, right? But this is probably nitpicking.

File: src/NautyInterface.jl [hunk 7]
```diff
@@ -0,0 +1,434 @@
...
+      set_subpart!(Y, inv_dict[src_ind], h, attrdict[t][inv_dict[tgt_ind]])
+    end
+  end
+  Y
+end
+
+"""
+Construct input for dreadnaut to compute automorphism group generators,
+canonical permutation/hash, and orbits.
+"""
```
[the remark this answers is not in the document]
Comment (KevinArlin) [reply src/NautyInterface.jl:340]: Sounds good to me.

File: src/NautyInterface.jl (hunk 1 above)
[the remark this answers is not in the document]
Comment (KevinArlin) [reply src/NautyInterface.jl:357]: Yeah, that's how it works for Julia arrays and suchlike. I'll post an issue.

File: src/NautyInterface.jl (hunk 2 above)
[the remark this answers is not in the document]
Comment (KevinArlin) [reply src/NautyInterface.jl]: I guess I don't mind as long as this is all getting fleshed out internally.

File: src/NautyInterface.jl [hunk 8]
```diff
@@ -0,0 +1,434 @@
...
+NautyResults must satisfy the following interface
+
+- strhsh: string value representing the isomorphism class
+- orbits: partitions of Cset parts into orbits
+  e.g. if E#2 = E#5, then these two elements are symmetric
+- generators: generating automorphisms of the automorphism group
+- ngroup: number of elements in the automorphism group
+- canonmap: isomorphism from the input into the canonoical isomorph
+- canon: canonical isomorph (codom of `canonmap`)
+"""
```
[the remark this answers is not in the document]
Comment (KevinArlin) [reply src/NautyInterface.jl:50]: Yep, sounds good.

**Review by KevinArlin — APPROVED** [review]

**Pull request merged**